doc/developer/design/20240625_subsource_deprecation_blue_green_sources.md

```diff
@@ -0,0 +1,317 @@
+inserts this into the appropriate `IngestionDescription` of the top-level
+source. This is used in source rendering to figure out what upstream
+reference should be mapped to a specific output collection.
+
+The storage `SourceExport` struct will be updated to include an optional
+projection field, populated from the `IngestionExport`'s `external_projection`:
+```rust
+pub struct SourceExport<O: proptest::prelude::Arbitrary, S = ()> {
+   /// Which output from the ingestion this source refers to.
+   pub ingestion_output: O,
+   /// The collection metadata needed to write the exported data
+   pub storage_metadata: S,
+   /// How to project the ingestion output to this export's collection.
+   pub output_projection: Option<Vec<u8>>,
+}
+```
+
+#### Source Rendering
+
+Currently, when an `IngestionDescription` is used inside `build_ingestion_dataflow`
+(the method that renders the various operators needed to run a source ingestion),
+it determines the `output_index` of each collection to be output by the ingestion
+by calling the `IngestionDescription::source_exports_with_output_indices` method.
+
+This method maps the `external_reference` (the upstream table name) of each `SourceExport`
+to the index of that upstream table in the source's `details` struct (which is created
+during purification -- its a vector of upstream table descriptions).
+
+Then inside each source implementation, we currently assume that each
+`SourceExport` uniqely corresponds to a single `output_index`, and that each `output_index`
+corresponds to the same index of the upstream table in the source's upstream `details`.
+
+This assumption is part of what prevents an upstream table from being ingested into
+more than one collection.
+
+We would update each source implementation to assume that more than one `SourceExport`
+can share an `output_index` and we would update the `source_exports` map created
+by `source_exports_with_output_indices` to include the `details` index explicitly.
+
+The `build_ingestion_dataflow` method currently demuxes the output collection of
+each source by each `output_index` and pushes each data stream to the appropriate
+`persist_sink`. We would update this method so that it applies the `output_projection`
+for each `SourceExport` on each data stream before being pushed to its `persist_sink`.
+
+#### Migration of source names (progress collections) & kafka source -> table migration
+
+TODO
+
+## Minimal Viable Prototype
+
+
+## Alternatives
+
+<!--
+What other solutions were considered, and why weren't they chosen?
+
+This is your chance to demonstrate that you've fully discovered the problem.
+Alternative solutions can come from many places, like: you or your Materialize
+team members, our customers, our prospects, academic research, prior art, or
+competitive research. One of our company values is to "do the reading" and
+to "write things down." This is your opportunity to demonstrate both!
+-->
+
+## Open questions
```

Comment:
Could you add a section around the system catalog, and how this affects existing source-specific introspection (_e.g._ what we report in `mz_source_statistics`)?

Reply:
As Roshan and I discussed this morning, this was the one thing I was wondering about too.